### ISSUE TITLE
curl does not use encoding of CLI for Basic Authorization on Windows

### ISSUE THREAD

**Issue by c-kunz** [curl/curl#6089]
# I did this

```bash
chcp 936
curl -v -u X:Ü https://...
```

### I expected the following

Authorization header contains Base64 encoded credentials in the encoding of the codepage 936:

`Authorization: Basic WDo=`

### I've got instead:

`Authorization: Basic WDrDnA==`

This seems to be `UTF-8` encoded instead, despite the command line using CP936.

### curl/libcurl version

`curl --version`:

```
curl 7.73.0 (x86_64-pc-win32) libcurl/7.73.0 OpenSSL/1.1.1h (Schannel) zlib/1.2.11 brotli/1.0.9 zstd/1.4.5 WinIDN libssh2/1.9.0 nghttp2/1.41.0
Release-Date: 2020-10-14
```

### operating system

Windows 10 Version 2004 (Build 19041.450)

Happens when curl is invoked within `cmd.exe` as well as `PowerShell`.

### Hint

On Ubuntu 20.04 `curl` seems to always immediately respect the encoding of the terminal, which is what I would also expect to happen on Windows.

My Ubuntu 20.04 curl is:

```
curl 7.68.0 (x86_64-pc-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3
Release-Date: 2020-01-08
```

**Comment by bagder** [comment]
curl just uses what it gets as is. There's no conversion or encoding involved.

**Comment by mkarg** [comment]
> curl just uses what it gets as is. There's no conversion or encoding involved.

Why does the conversion happen on Ubunto then?

**Comment by c-kunz** [comment]
Indeed, I found out yesterday that curl in the following version on the same Windows machine:

```
curl 7.55.1 (Windows) libcurl/7.55.1 WinSSL
Release-Date: 2017-11-14, security patched: 2019-11-05
```

is using `ISO-8859-1` (CP850) to encode the credentials. This got me confused as well when upgrading to the latest curl which suddenly uses UTF-8, but nonetheless never respects the active codepage of the CLI.

**Comment by MarcelRaad** [comment]
Probably caused by https://github.com/curl/curl/commit/9e5669f3880674de262000e772ef21ec0cc874be.

**Comment by ChuVienTran** [comment]
Good

**Comment by ChuVienTran** [comment]
Windo

**Comment by TvdW** [comment]
Stumbled across this as I occasionally crawl through the curl issue tracker, and found myself wanting to leave a comment on this one.

[RFC 7617](https://tools.ietf.org/html/rfc7617) (2015) added a 'charset' parameter to the authentication challenge, and specifies that only UTF-8 is a valid argument, so that the internet could have a migration path out of the previously existing mess of having every charset supported with servers having to use heuristics to figure out what the client is sending. Prior to that RFC the charset was simply undefined but most servers used ISO-8859-1 and so clients did as well. If the new parameter is unset the charset is still undefined, if it is set the charset is defined to be UTF-8. In practice means that the encoding is either UTF-8 (when the server sets the parameter) or US-ASCII (which is conveniently the set of characters the two have in common).

Nowadays the mainstream browsers use UTF-8 for encoding the basic authentication header. Ironically Mozilla's [Bugzilla entry](https://bugzilla.mozilla.org/show_bug.cgi?id=1419658) on the topic refers to curl, and states that `At this point in time though, the rest of the ecosystem (Chrome, curl, nginx, and likely others) have settled on utf8.` (apparently not!)

Ultimately you're hitting a piece of undefined behavior, in which the client has to make the call on which charset to use. I personally don't think that the terminal charset should influence anything other than how stdin/stdout are encoded, so curl's behavior here looks correct to me, even if it is by accident. RFC 7617 does not specify a default charset when the parameter is left out, but does hint at a future in which everything is UTF-8. That was five years ago, and browsers have made the switch, thus I would conclude that the future is now and this is not a bug, but a feature. 😉

**Comment by mkarg** [comment]
Well, *someone* must decide about the actual charset to use:
* cURL could use the bytes as given by the terminal.
* cURL could have an optional encoding option.
* cURL could send *two* requests using an unauthenticated first, followed by one that respects the charset responeded by the server.
But *all* of these choices imply that the cURL team has to decide and to actively code them.

**Comment by bagder** [comment]
> cURL could use the bytes as given by the terminal

Except for Windows (as mentioned above) this is what curl already does.

> cURL could have an optional encoding option.

I would really dislike that as character encoding/decoding is a total mess and requires a lot of translation tables etc.

> cURL could send two requests using an unauthenticated first

It really couldn't without the user explicitly *asking* for that behavior as it seems rather intrusive with a risk for weird side-effects.

**Comment by mkarg** [comment]
Well, we go in circles here, as this issue solely *is* about Windows...

**Comment by bagder** [comment]
Sorry, of course it is. But I don't know what the answer is for Windows so I'll shut up.

**Comment by vszakats** [comment]
Decoding `WDo=` is `X:`, in this case meaning no password instead of the passed `Ü`. This is not necessarily a result expected universally. If here is something to fix here, is platforms that are at the moment passing around these strings as raw byte-streams, and therefore require to have all the non-curl, client/server components to have the same interpretration of those streams. Switching these platform to use UTF-8 internally, would have a long-term, universal benefit IMO.

The official curl Windows binaries also have Unicode enabled, but it's possible to create local builds without it. Speaking of `curl-for-win`, [this](https://github.com/curl/curl-for-win/blob/9e308246fa552cb494673c475059912eb01ff289/curl.sh#L65) line needs to deleted for the effect.

**Comment by xquery** [comment]
an unclear area for sure ... now we have documented can this issue be closed ?

**Comment by mkarg** [comment]
I do not see why we should close this issue. There is neither a solution nor a workaround. Hence it must stay open.

**Comment by xquery** [comment]
sorry, my comment was not directed at the intractable nature of this issue.

let me try again - we have a page documenting known bugs here:

  https://curl.se/docs/knownbugs.html

we do not keep open issues for each these in the issue tracker ... as I did not see anyone in this thread saying they would take on any work (perhaps I missed it ?) I suggested as we documented it that this issue can be closed.

**Comment by vszakats** [comment]
I believe curl is working correctly here by reading values passed over the command-line as Unicode strings, presented by Windows as-is. Windows is the only component that can determine how to interpret various codepoints passed to programs.

The bug I see here is that this curl feature is not yet consistently enabled by default across all build-systems. But, if enabled consistently, it will still not give the result expected by this report.

If an invalid/undefined value is expected for compatibiliy with certain configurations/platforms, I think the only way to solve this is to add the header with a manually calculated base64 string from whatever binary value that is expected.

**Comment by mkarg** [comment]
@vszakats You misunderstood the situation. Windows is *not* passing UTF-8, it is passing bytes in the locally enabled code page (see the actual behavior described in the original bug report). Hence it is *curl* which translates it into UTF-8 internally, which is a *bug*, as you all said that curl *would not* do that conversion!

**Comment by xquery** [comment]
Yes @mkarg we have documented as a known bug. The last discussion was ~6 mths ago and wanted to check if you or anyone is planning a PR ?

**Comment by mkarg** [comment]
I am too involved in other open source projects so I think I cannot provide an PR any time soon.

**Comment by bagder** [comment]
If nobody works on the issue and it's been open for many months (and this issue qualifies), then we should detail it in the `KNOWN_BUGS` document and close this issue on GitHub. We don't leave old stalled bugs open.

**Comment by vszakats** [comment]
@mkarg Can you pinpoint the curl source code which does the codepage conversion in this particular case?

**Comment by mkarg** [comment]
Not today. I am not a curl contributor and first have to dive into the code to learn how it works. What I do know is that Windows definitively doesn not send UTF-8 until you explicitly ask it to.

**Comment by vszakats** [comment]
Somebody correct if this is wrong, but it appears as if `curlx_convert_wchar_to_UTF8()` is the only function doing conversion here, where Windows is passing a Unicode `wchar` string (encoded as UTF-16) to curl which curl then losslessly re-encodes as UTF-8 internally (via `WideCharToMultiByte()`, to be passed over the network. So, unless curl gets a new feature where it can be instructed to interpret Unicode text input in a specific codepage/encoding before sending over the wire, I can't see how anything more can be done here. Even if curl gets such feature, the above example will still _not_ work as expected in the original post, as it will then need whatever codepage value passed to it directly (via a new option) and not via `chcp` or other OS means.

So, my understanding is that the expectation reported here can only be fixed by restoring one specific undefined behavior, while breaking well-defined behaviour and breaking other, undefined cases of 8-bit to 8-bit conversion.

**Comment by bagder** [comment]
> I can't see how anything more can be done here

curl doesn't convert the provided input - by design. The user needs to provide the data in the correct format.
